[AD] Indirect passing differentiation support.

dan-zheng opened the pull request:
This patch adds initial support for differentiation of functions with indirect parameters/results.
There are no regressions for non-indirect-passing differentiation.
Indirect passing TODOs/FIXMEs are tracked at the primary bug [TF-104](https://bugs.swift.org/browse/TF-104).

- Implement reabstraction thunk generation helpers, adapting code from SILGen.
  - `getOrCreateReabstractionThunk` creates a thunk given source and target function types, handling arbitrary parameter/result reabstraction.
- Support indirect passing in PrimalGen.
  - In `PrimalGenCloner::visitApplyInst()`:
    - Reabstract pullbacks using thunks.
      - This is necessary when actual pullback types are "less direct" than the lowered
        primal value struct pullback member types, which are "as direct as possible".
  - In `PrimalGenCloner::getOrDiagnoseAssociatedFunctionSubstitutionMap`:
    - Check callee VJP same-type generic requirements.
      Previously, only conformance requirements were checked.
- Support indirect passing in AdjointGen.
  - In `AdjointEmitter::run()`:
    - Modularly register adjoint buffers for original indirect parameters and indirect seeds.
  - In `AdjointEmitter::visitApplyInst()`:
    - Select original `apply` result based on source index.
    - Reabstract pullbacks using thunks, in symmetry with PrimalGen.
    - Modularly create stack allocations for pullback indirect results.
    - Handle adjoint buffer accumulation for original `apply` indirect parameters.
  - Implement `AdjointEmitter::visitCopyAddr`.
  - Fix cleanup/deallocation for adjoint buffers.
     - Use `(retain|release)_value_addr` only for buffers with loadable types.
       Otherwise, cleanup using `destroy_addr`.
- Random fixes/gardening.
  - Fix generic signature handling in `SILFunctionType::getAutoDiffAssociatedFunctionType`.
    - Canonicalize result types using generic signature, if specified.
      This is important for consistency to avoid deserialization verification crashes.
  - Check callee VJP same-type generic requirements during PrimalGen.
    - This requires additional logic in `PrimalGenCloner::getOrDiagnoseAssociatedFunctionSubstitutionMap`.
- Add new indirect passing test files to avoid affecting existing tests.
  - `test/TensorFlowRuntime/tensor_autodiff_indirect.swift`: currently crashes with `-O` ([TF-199](https://bugs.swift.org/browse/TF-199)).
  - `test/AutoDiff/autodiff_indirect_diagnostics.swift`: currently requires `-verify-ignore-unknown` ([TF-201](https://bugs.swift.org/browse/TF-201)).

dan-zheng committed:
[AutoDiff] Start indirect passing differentiation support.

This patch enables differentiation of functions with indirect parameters/results.

- Change differentiation pass to handle function calls with indirect parameters/results.
- Handle pullback abstraction mismatches, if necessary.
  - Generate/call reabstraction thunks before storing pullback in primal values struct.
- Use `(retain|release)_value_addr` only for buffers with loadable types.

Todos before merging (in order of priority):
- Fix `AdjointEmitter::visitApplyInst`.
  - FIXME comments mark crucial code to fix.
  - AD crashes/miscomputes for some programs.
    - See tests `Indirect passing (4|5)` in `test/TensorFlowRuntime/tensor_autodiff_runtime.swift`.
- Diagnose differentiation crashes with `-O`.
  - `test/TensorFlowRuntime/tensor_autodiff_runtime.swift` crashes with `-O`.
- Diagnose `no conformance to 'AdditiveArithmetic'` due to generic minimization bug.
- Diagnose direct differentiation of reabstraction thunks.
  - Handling this case requires more work.
  - The place to diagnose is marked in `emitAssociatedFunctionReference`.
- Clean up thunk generation code.

dan-zheng committed:
[AutoDiff] Finish initial indirect passing support.

Known derivative miscomputations related to indirect passing have all been fixed.
There are no regressions for non-indirect-passing differentiation.

This patch is ready to merged (as an initial patch for indirect passing).
TODOs/FIXMEs are tracked at: TF-199, TF-201, TF-202, TF-204.

- Fix `AdjointEmitter::visitApplyInst`.
  - Adjoint buffers for original indirect parameters are now correctly
    initialized and passed to pullbacks if necessary.
- Fix generic signature handling `getAutoDiffAssociatedFunctionType`.
  - Result types should be canonicalized using generic signature,
    if specified. This is important for consistency to avoid deserialization
    verification crashes.
- Check callee VJP same-type generic requirements during PrimalGen.
  - This requires additional logic in
    `PrimalGenCloner::getOrDiagnoseAssociatedFunctionSubstitutionMap`.
- Refactor reabstraction thunk generation helpers.
  - Thunk helpers are now all top-level `static` functions.
- Add tests.

dan-zheng commented:
@rxwei: Ready for review. Triggering tests to make sure everything passes as expected.

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -3397,45 +3810,114 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
-    materializeAdjointIndirect(std::move(av), seedBuf);
-    if (av.getType().isAddressOnly(getModule()))
+    if (adjointType.isObject()) {
+      ValueWithCleanup access(
```
Why do you need a `ValueWithCleanup` here? I'm only seeing builder calls in this scope, which just take a `SILValue`.

dan-zheng replied:
Sorry, I blindly copied `ValueWithCleanup access` from another function.
Changed to just `auto access = builder.createBeginAccess(..)` in https://github.com/apple/swift/pull/22544/commits/c1ff15405abb1df3ec85d58a0b40df718376a90f.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -3397,45 +3810,114 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
+    }
+
+    // Initialize seed buffer.
+    ValueWithCleanup seedBuf(builder.createAllocStack(loc, adjointType),
```
I don't think `ValueWithCleanup` is necessary here. It's not being used by any calls below that take `ValueWithCleanup`.

dan-zheng replied:
Changed `seedBuf` to just `SILValue` in https://github.com/apple/swift/pull/22544/commits/c1ff15405abb1df3ec85d58a0b40df718376a90f.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -1815,6 +1811,12 @@ emitAssociatedFunctionReference(ADContext &context, SILBuilder &builder,
+    // Tentative solution: clone a new reabstraction thunk where function
+    // argument has a `@differentiable` function type.
+    if (originalFn->isThunk() == IsReabstractionThunk) {
+      // Handle here!
```
A diagnostic is already emitted when differentiating the reabstraction thunk because the thunk argument function is opaque:
```
[AD] Activity info for $sS2fIegnr_S2fIegyd_TR at (source=0 parameters=(0))
[ACTIVE] %0 = argument of bb0 : $Float                     // user: %3
[NONE] %1 = argument of bb0 : $@callee_guaranteed (@in_guaranteed Float) -> @out Float // user: %5
[ACTIVE]   %2 = alloc_stack $Float                         // users: %3, %8, %5
[ACTIVE]   %4 = alloc_stack $Float                         // users: %6, %7, %5
[VARIED]   %5 = apply %1(%4, %2) : $@callee_guaranteed (@in_guaranteed Float) -> @out Float
[ACTIVE]   %6 = load %4 : $*Float                          // user: %9
[AD] Cloning original @$sS2fIegnr_S2fIegyd_TR to primal @AD__$sS2fIegnr_S2fIegyd_TR__primal_src_0_wrt_0
[AD] Primal-transforming:
  %5 = apply %1(%4, %2) : $@callee_guaranteed (@in_guaranteed Float) -> @out Float

[AD] Active indices: params={0}, results={0}
<unknown>:0: note: opaque non-'@differentiable' function is not differentiable
[AD] Diagnosing non-differentiability for inst
	  %5 = apply %1(%4, %2) : $@callee_guaranteed (@in_guaranteed Float) -> @out Float

while performing differentiation task
	(differentiation_task original=@$sS2fIegnr_S2fIegyd_TR attribute=source 0 wrt 0 jvp @AD__$sS2fIegnr_S2fIegyd_TR__jvp_src_0_wrt_0 vjp @AD__$sS2fIegnr_S2fIegyd_TR__vjp_src_0_wrt_0 invoker=(differentiation_invoker differential_operator=((autodiff_function implicit type='@differentiable (Float) -> Float' location=ind2.swift:7:34 range=[ind2.swift:7:34 - line:7:34]
  (function_conversion_expr implicit type='(Float) -> Float' location=ind2.swift:7:34 range=[ind2.swift:7:34 - line:7:34]
    (declref_expr type='(Float) -> Float' location=ind2.swift:7:34 range=[ind2.swift:7:34 - line:7:34] decl=ind2.(file).direct@ind2.swift:3:6 [with (substitution_map generic_signature=<T where T : Differentiable, T.CotangentVector : AdditiveArithmetic, T.TangentVector : AdditiveArithmetic> (substitution T -> Float))] function_ref=unapplied))))))
ind2.swift:7:34: error: function is not differentiable
print(pullback(at: Float(1), in: direct)(1))
                                 ^~~~~~
<unknown>:0: note: expression is not differentiable
```

I mainly marked this spot so that "direct differentiation of indirect functions" can be handled in a follow-up (addressing TF-201). Let me know if you feel like we should explicitly emit an error here for now.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -3338,17 +3736,32 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
       assert(useIt != alloc->use_end());
       SILBasicBlock *deallocationPoint = (useIt++)->getUser()->getParent();
-      do {
+      while (useIt != alloc->use_end()) {
```
We discussed in person: the new code is intentional and fine.

The old code always increments `useIt` twice, though it is not always the case that `alloc` has two uses.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -3397,45 +3810,114 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
+      adjointType = adjointBuf.getValue()->getType();
+    }
+
+    // Register indirect results as arguments.
```
Can you elaborate a bit more on what this is doing? I don't understand what `tempAllocation` is for.

dan-zheng replied:
Wow, the original code was completely wrong.

Things are fixed and made clear in https://github.com/apple/swift/pull/22544/commits/6f068338919019d8ae7d1178900d5ada0d050c7d.
The code here in `AdjointEmitter::visitApplyInst` allocates buffers for pullback indirect results. Those indirect result buffers may then be accumulated into original indirect parameters' adjoint buffers, depending on the differentiation parameter indices.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -3741,8 +4259,12 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
+              else
+                b.createDestroyAddr(l, v);
             })));
+    // localAllocations.push_back(adjBuf);
```
Remove.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/22544/commits/c1ff15405abb1df3ec85d58a0b40df718376a90f.

dan-zheng committed:
Minor edits.

- Address feedback from @rxwei.
- Todo: clean up code surrounding `tempAllocations`.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -3397,45 +3810,114 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
                                              /*isNonThrowing*/ false);
-    builder.createDeallocStack(loc, seedBuf);
+    tempAllocations.push_back(seedBuf.getValue());
+    // builder.createDeallocStack(loc, seedBuf);
```
Removed in https://github.com/apple/swift/pull/22544/commits/c1ff15405abb1df3ec85d58a0b40df718376a90f.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -3397,45 +3810,114 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
+    } else {
+      auto adjointBuf = getAdjointBuffer(origResult);
+      adjointValue = adjointBuf.getValue();
+      adjointType = adjointBuf.getValue()->getType();
```
Done in https://github.com/apple/swift/pull/22544/commits/c1ff15405abb1df3ec85d58a0b40df718376a90f.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -3397,45 +3810,114 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
-    SILValue pullback = builder.createStructExtract(loc,
-                                                    primalValueAggregateInAdj,
-                                                    field);
+    SILValue pullback =
```
Done in https://github.com/apple/swift/pull/22544/commits/c1ff15405abb1df3ec85d58a0b40df718376a90f.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -3320,13 +3710,21 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
+        if (auto *cleanup = adjBuf.getCleanup()) {
+          LLVM_DEBUG(getADDebugStream() << "Disabling cleanup for "
+                     << adjBuf.getValue() << "for return\n");
+          cleanup->disable();
```
Done in https://github.com/apple/swift/pull/22544/commits/c1ff15405abb1df3ec85d58a0b40df718376a90f.

dan-zheng committed:
Fix `AdjointEmitter::visitApplyInst`.

- Fix code for creating allocations for pullback indirect results.
- Reorganize code, improve doc comments.

dan-zheng committed:
[AutoDiff] Reabstract pullback in adjoint for symmetry with primal.

- Reabstract pullback in adjoint for symmetry with primal.
  - Store original pullback type in `NestedApplyInfo` before reabstraction during PrimalGen.
  - If pullback is reabstracted in primal, reabstract pullback struct member in adjoint.
- Generalize `createReabstractionThunk` to handle arbitrary parameter/result reabstraction.
  - Previously, only indirect->direct reabstraction was handled.
- Significantly cleanup `AdjointEmitter::visitApplyInst`.
- Minor gardening.

Addresses feedback from @rxwei.

dan-zheng committed:
Merge branch 'tensorflow-real' into ad-indirect-passing

dan-zheng committed:
Minor edits.

- Rename `createReabstractionThunk` to `getOrCreateReabstractionThunk`.
- Remove unnecessary `tempAllocations` array for tracking local allocations.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -2494,7 +2836,39 @@ class PrimalGenCloner final
+            .castTo<SILFunctionType>();
+    if (!loweredPullbackType->isEqual(actualPullbackType)) {
+      // Set original pullback type before reabstraction.
+      // NOTE: Is there a better `DenseMap` API?
```
@rxwei Is there a shorter `DenseMap` API than this super long call here?
I just want to update the `NestedApplyInfo.originalPullbackType` field.

rxwei replied:
Answered in a separate review comment because i didn't see this.

dan-zheng committed:
Minor edits.

- Debug print for reabstraction thunk creation.
- Gardening.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -3793,11 +4300,38 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
+    addToAdjointBuffer(cai->getSrc(), readAccess);
+    builder.createEndAccess(cai->getLoc(), readAccess, /*aborted*/ false);
+    auto destType = remapType(adjDest.getType());
+    emitZeroIndirect(destType.getASTType(), adjDest, cai->getLoc());
```
`emitZeroIndirect` takes an access. You need `begin_access [modify]` and `end_access`.

rxwei replied:
Zero needs a new cleanup. You should do what `visitStoreInst` is doing, including matching access kinds.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/22544/commits/ad7436a8e38562bbf4aa6cf6ebe6918c829c5331.

rxwei commented on test/TensorFlowRuntime/tensor_autodiff_runtime.swift:
```diff
@@ -4,7 +4,9 @@
 // REQUIRES: executable_test
-// REQUIRES: swift_test_mode_optimize
+//
+// FIXME(TF-199): Indirect passing differentiation crashes with `-O`.
```
We shouldn't disable `-O` for existing tests. Could you create a new file for testing indirect passing?

dan-zheng replied:
Very good point.
Created new `tensor_autodiff_indirect.swift` test in https://github.com/apple/swift/pull/22544/commits/ad7436a8e38562bbf4aa6cf6ebe6918c829c5331. `-O` is re-supported for `tensor_autodiff_runtime.swift`.

rxwei commented:
Before `-O` is fixed, I think we should make indirect passing support available under a flag.

rxwei commented:
I'd suggest playing with [DeepLearning](http://github.com/tensorflow/swift-api) to see whether things work for libraries.

rxwei commented:
Lastly, congrats on figuring everything out! This is a huge huge step forward.

dan-zheng committed:
Address review comments, separate indirect passing differentiation tests.

- Address review comments from @rxwei.
- Create separate indirect passing differentiation tests so that original tests
  aren't affected by workarounds.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -2494,7 +2836,39 @@ class PrimalGenCloner final
+            .castTo<SILFunctionType>();
+    if (!loweredPullbackType->isEqual(actualPullbackType)) {
+      // Set original pullback type before reabstraction.
+      // NOTE: Is there a better `DenseMap` API?
```
We discussed in-person, thank you for the detailed walkthrough!
Changed to use `try_emplace` in https://github.com/apple/swift/pull/22544/commits/ad7436a8e38562bbf4aa6cf6ebe6918c829c5331. We agreed that set-and-modify is actually more elegant using `try_emplace`.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -2446,30 +2797,21 @@ class PrimalGenCloner final
+    // Record desired/actual VJP indices.
+    // Temporarily set original pullback type to `None`.
+    getDifferentiationTask()->getNestedApplyInfo().insert(
+        {ai, {indices, vjpAndVJPIndices->second, None}});
```
Resolved in https://github.com/apple/swift/pull/22544/commits/ad7436a8e38562bbf4aa6cf6ebe6918c829c5331.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -3246,6 +3641,35 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
                << "Assigned seed " << *seed << " as the adjoint of "
                << formalResults[srcIdx]);
 
+    // Pre-emptively register adjoint buffers for original indirect parameters.
```
`register` -> `set` in https://github.com/apple/swift/pull/22544/commits/ad7436a8e38562bbf4aa6cf6ebe6918c829c5331.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -1928,6 +1932,317 @@ emitAssociatedFunctionReference(ADContext &context, SILBuilder &builder,
+
+/// Get or create a reabstraction thunk from `fromType` to `toType`, to be
+/// called in `caller`.
+SILFunction *getOrCreateReabstractionThunk(SILOptFunctionBuilder fb,
```
static

dan-zheng replied:
Done in https://github.com/apple/swift/pull/22544/commits/53701023ac08031c3c267b709072b6f137d9de4d.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -2308,7 +2623,8 @@ class PrimalGenCloner final
     SILValue pullback = vjpDirectResults.back();
-    getPrimalInfo().addPullbackDecl(sei, pullback->getType());
+    // TODO: Check whether it's necessary to reabstract getter pullbacks.
+    getPrimalInfo().addPullbackDecl(sei, getOpType(pullback->getType()));
```
It is totally necessary. You can write a computed/stored property that returns a generic parameter in the parent type to test it. Write a custom VJP to avoid triggering `struct_element_addr` differentiation.

dan-zheng replied:
I tried to write tests that would trigger this code path (requiring getter pullback reabstraction) but failed. Initial experiments:

```swift
protocol Addable : Differentiable & FloatingPoint {
  @differentiable(wrt: (x, y))
  static func add(_ x: Self, _ y: Self) -> Self
}

struct Foo<Scalar : Differentiable> : Differentiable
  // where Scalar : AdditiveArithmetic, Scalar.TangentVector : AdditiveArithmetic, Scalar.CotangentVector : AdditiveArithmetic
  where Scalar : Addable, Scalar.TangentVector : AdditiveArithmetic, Scalar.CotangentVector : AdditiveArithmetic
{
  var member: Scalar

  var computed: Scalar {
    // return .zero + member
    return Scalar.add(.zero, member)
  }
}

func test<Scalar : Differentiable>(_ x: Foo<Scalar>) {
  print(pullback(at: Foo<Scalar>(member: Scalar.zero), in: { foo in foo.computed }))
}
```

I think we can implement getter pullback reabstraction lazily, once we have an effective test.

dan-zheng committed:
Minor edits.
- Make `getOrCreateReabstractionThunk` static.
- Reference bug TF-210 in TODO comment.

Address feedback from @rxwei.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -3499,13 +3986,30 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
     }
+    // Deallocate pullback indirect results.
+    for (auto *alloc : reversed(pullbackIndirectResults))
+      builder.createDeallocStack(loc, alloc);
```
Done in https://github.com/apple/swift/pull/22544/commits/ad7436a8e38562bbf4aa6cf6ebe6918c829c5331.
Added TODO comment in https://github.com/apple/swift/pull/22544/commits/53701023ac08031c3c267b709072b6f137d9de4d referencing TF-210.

dan-zheng commented:
> Before `-O` is fixed, I think we should make indirect passing support available under a flag.

I tried to gate indirect passing support via a flag: WIP branch at [dan-zheng/swift:ad-indirect-passing-flag](https://github.com/dan-zheng/swift/tree/ad-indirect-passing-flag).

The blocker is `test/TensorFlowRuntime/tensor_autodiff_indirect.swift`, for which the following `RUN` line doesn't work:
```swift
// FIXME: Passing arbitrary arguments after `%target-run-eager-swift` doesn't work.
// RUN: %target-run-eager-swift -Xllvm -indirect-passing-differentiation
```

---

We discussed in person and decided that not adding a flag is okay.
* Indirect passing is known to crash with `-O` only for niche cases (`"ResultSelection"` test in `test/TensorFlowRuntime/tensor_autodiff_indirect.swift`).
* Unsupported indirect passing cases don't crash - diagnostics are emitted, which is acceptable. ([TF-201](https://bugs.swift.org/browse/TF-201), [TF-203](https://bugs.swift.org/browse/TF-203)).

rxwei commented:
LGTM!

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
Merged to unblock progress. Known indirect passing issues are tracked at primary bug [TF-104](https://bugs.swift.org/browse/TF-104).
